File: app/src/main/java/com/example/roomdatabase/MainActivity.java
package com.example.roomdatabase;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    EditText userId, password, name;
    Button register;
    Button login;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        userId = findViewById(R.id.userId);
        password = findViewById(R.id.password);
        name = findViewById(R.id.name);
        register = findViewById(R.id.register);
        login = findViewById(R.id.login);

                register.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                UserEntity userEntity = new UserEntity();
                userEntity.setUserId(userId.getText().toString());
                userEntity.setPassword(password.getText().toString());
                userEntity.setName(name.getText().toString());
                if (validateInput(userEntity)) {
                        UserDatabase userDatabase= UserDatabase.getUserDatabase(getApplicationContext());
                        UserDao userDao  = userDatabase.userDao();
                        new Thread(new Runnable() {
                            @Override
                            public void run() {
                                //register User
                                userDao.registerUser(userEntity);
                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        Toast.makeText(getApplicationContext(), "User Register", Toast.LENGTH_SHORT).show();

                                    }
                                });

                            }
                        }).start();
                } else {
                    Toast.makeText(getApplicationContext(), "Fill all fields",Toast.LENGTH_SHORT).show();
                }

            }
        });

        login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
            startActivities(new Intent[]{new Intent(MainActivity.this, Login.class)});
            }
        });
    }
    private Boolean validateInput(UserEntity userEntity){
        if (userEntity.getName().isEmpty()||
        userEntity.getPassword().isEmpty()||
        userEntity.getName().isEmpty()){
            return false;
        }
        return true;
    }

}

File: app/src/main/java/com/example/roomdatabase/MainActivity.kt
package com.example.roomdatabase

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.roomdatabase.databinding.ActivityMainBinding
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var firebaseDatabase: FirebaseDatabase
    private lateinit var databaseReference: DatabaseReference


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        firebaseDatabase = FirebaseDatabase.getInstance()
        databaseReference= firebaseDatabase.reference.child("users")

        binding.signupButton.setOnClickListener{
            val signupUsername = binding.signupUsername.text.toString()
            val signupPassword = binding.signupPassword.text.toString()
            if (signupUsername.isNotEmpty()&& signupPassword.isNotEmpty()){
                mainUser(signupUsername, signupPassword)
            } else{
                Toast.makeText(this@MainActivity, "All fields are mandatory ", Toast.LENGTH_SHORT).show()
            }

        }
        binding.loginRedirect.setOnClickListener {
            startActivity(Intent(this@MainActivity, Login::class.java))
            finish()
        }


    }
    private fun mainUser(username: String, password: String){
        databaseReference.orderByChild("username").equalTo(username).addListenerForSingleValueEvent(object : ValueEventListener{
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (!dataSnapshot.exists()){
                    val id = databaseReference.push().key
                    val userData = UserData(id, username,password)
                    databaseReference.child(id!!).setValue(userData)
                    Toast.makeText(this@MainActivity, "Signup Successful", Toast.LENGTH_SHORT).show()
                    startActivity(Intent(this@MainActivity, Login::class.java))
                    finish()
                } else{
                    Toast.makeText(this@MainActivity, "User already exists ", Toast.LENGTH_SHORT).show()

                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                Toast.makeText(this@MainActivity, "Database Error: ${databaseError.message}", Toast.LENGTH_SHORT).show()

            }
        })
    }

}
